functorch/functorch/csrc/BatchRulesRandomness.cpp

```diff
@@ -227,15 +227,16 @@ Tensor multinomial_batching_rule(const Tensor& self, const int64_t num_samples,
     shapeVec.insert(shapeVec.end(), shape.begin(), shape.end());
     self_value = self_value.expand(shapeVec);
   }
-  if (self_value.dim() == 3 && (self_bdim || randomness == RandomnessType::Different)) {
+  const auto orig_input_dim = self_value.dim();
+  if (orig_input_dim == 3 && (self_bdim || randomness == RandomnessType::Different)) {
```

Comment:
nit: naming a bit off-putting. `orig_input_dim` -- `self_value` on line 218 may have a different shape from self_value on line 228.

Also, this implementation of `multinomial` is confusing, since we're modifying it anyways it is a good chance to clean it up. The thing that is confusing is that it's hard to tell what actually happens if randomness is Same vs Different and what happens in the 1D or 2D multinomial cases.

It might be clearer if we just separate the cases and use helper functions like `rankWithoutBatchDim` to tell us if we're in the 1D or 2D case and `ensure_has_bdim` instead of the manual expand logic above. Pseudocode:
```cpp
if (randomness == RandomnessType::Different) {
  // 1D cases: S -> BS -> multinomial(BS)
  //          BS -> multinomial(BS)
  // 2d cases: MS -> (BM)S -> multinomial(BS)
  //.         BMS -> (BM)S -> multinomial(BS)
  is_2d_case = rankWithoutBatchDim(self_value, self_bdim);
  if (self_bdim) {
    self_value = ensure_has_bdim(self_value, ...)
    if (is_2d_case) {
        self_value = reshape_dim_into(0, 0, self_value);
    }
  }
  auto out = multinomial(...);
  if (is_2d_case) {
    out = reshape_dim_outof(...);
  }
}
if (randomness == RandomnessType::Same) {
  // 1D case: S -> multinomial(S)
  // 2D case: MS -> multinomial(MS)
  out = multinomial(self_value);
  return out;
}


```

Reply:
Good call--updated to match